### ISSUE TITLE
`renderPlot()` does not work with `bindEvent()`

### ISSUE THREAD

**Issue by kennedymwavu** [rstudio/shiny#3876]
### Problem description

When I bind `renderPlot()` to a trigger button using `bindEvent()`, the `renderPlot()` expression is only ran once (on the first click). Subsequent clicks on the button have no effect ie. plot is not re-rendered.

### Reprex

```R
library(shiny)

ui <- fluidPage(
  actionButton(inputId = "trigger", label = "Plot"),
  plotOutput("myPlot")
)

server <- function(input, output, session) {
  output$myPlot <- renderPlot({
    cols <- sample(names(iris)[1:4], size = 2, replace = FALSE)
    cat("`renderPlot()` expr executed.\n")
    plot(
      iris[, cols],
      col = iris[["Species"]],
      pch = 19
    )
  }) |> bindEvent(input$trigger)
  
  # initialize button click counter:
  counter <- reactiveVal(0L)
  observeEvent(input$trigger, {
    counter(counter() + 1)
    # print number of button clicks:
    cat("Clicks: ", counter(), "\n")
  })
}

shinyApp(ui, server)
```

Might this be a possible bug or am I missing something?

### More details

1. The same reprex above will work as expected if I put the `renderPlot()` inside `observeEvent()`.

<details>
<summary>
Show/hide the reprex using <code>observeEvent()</code>
</summary>

```R

library(shiny)

ui <- fluidPage(
  actionButton(inputId = "trigger", label = "Plot"),
  plotOutput("myPlot")
)

server <- function(input, output, session) {
  # initialize button click counter:
  counter <- reactiveVal(0L)
  observeEvent(input$trigger, {
    counter(counter() + 1)
    # print number of button clicks:
    cat("Clicks: ", counter(), "\n")
    
    output$myPlot <- renderPlot({
      cols <- sample(names(iris)[1:4], size = 2, replace = FALSE)
      cat("`renderPlot()` expr executed.\n")
      plot(
        iris[, cols],
        col = iris[["Species"]],
        pch = 19
      )
    })
  })
}

shinyApp(ui, server)
```

</details>

2. For comparison purposes, `renderTable()` works fine with `bindEvent()`, with the table output changing each time the button is clicked.

<details>
  <summary>Show/hide <code>renderTable()</code> reprex</summary>
  
```R
library(shiny)

ui <- fluidPage(
  actionButton(inputId = "trigger", label = "Trigger"),
  tableOutput(outputId = "tbl")
)

server <- function(input, output, session) {
  output$tbl <- renderTable({
    cols <- sample(x = names(iris)[1:4], size = 2, replace = FALSE)
    cat("`renderTable()` expr executed.\n")
    iris[, cols] |> head()
  }) |> bindEvent(input$trigger)
  
  # initialize button click counter:
  counter <- reactiveVal(0L)
  observeEvent(input$trigger, {
    counter(counter() + 1)
    # print number of button clicks:
    cat("Clicks: ", counter(), "\n")
  })
}

shinyApp(ui, server)
```
</details>

### System details

<details>
<summary>
Show/hide output of <code>sessionInfo()</code>
</summary>

```
R version 4.3.1 (2023-06-16)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.3 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0 
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Africa/Nairobi
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] shiny_1.7.5

loaded via a namespace (and not attached):
 [1] digest_0.6.33     later_1.3.1       R6_2.5.1          httpuv_1.6.11     fastmap_1.1.1    
 [6] magrittr_2.0.3    cachem_1.0.8      memoise_2.0.1     htmltools_0.5.6   lifecycle_1.0.3  
[11] promises_1.2.1    cli_3.6.1         xtable_1.8-4      sass_0.4.7        textshaping_0.3.6
[16] jquerylib_0.1.4   systemfonts_1.0.4 compiler_4.3.1    rstudioapi_0.15.0 tools_4.3.1      
[21] ragg_1.2.5        bslib_0.5.1       mime_0.12         ellipsis_0.3.2    Rcpp_1.0.11      
[26] jsonlite_1.8.7    rlang_1.1.1 

```

</details>

**Comment by gadenbuie** [comment]
Thanks for the excellent reprex and bug report! I agree, this seems like it might be a bug in how `renderPlot()` and `bindEvent()` work together. I would certainly have expected your plot to update.

Here's another example where essentially everything on the page should update, including the plot, but only the doesn't update.

<details>
<summary>Code</summary>

```r
library(shiny)

inline_text_output <- function(id) {
  textOutput(id, container = function(...) span(..., .noWS = "outside"))
}

ui <- fixedPage(
  h2("Issue 3876"),
  actionButton("update", "Update"),
  p(
    "Random number ", inline_text_output("clicks"), ": ",
    inline_text_output("random"), "."
  ),
  plotOutput("plot"),
  uiOutput("ui"),
  verbatimTextOutput("code")
)

server <- function(input, output, session) {
  output$clicks <-
    renderText({ input$update }) |>
    bindEvent(input$update, ignoreNULL = FALSE, ignoreInit = FALSE)

  output$random <-
    renderText({ runif(1) }) |>
    bindEvent(input$update, ignoreNULL = FALSE, ignoreInit = FALSE)

  output$plot <-
    renderPlot({ hist(runif(1000)) }) |>
    bindEvent(input$update, ignoreNULL = FALSE, ignoreInit = FALSE)

  output$ui <-
    renderUI({ lorem::ipsum(1, 1) }) |>
    bindEvent(input$update, ignoreNULL = FALSE, ignoreInit = FALSE)

  output$code <-
    renderPrint({ runif(3) }) |>
    bindEvent(input$update, ignoreNULL = FALSE, ignoreInit = FALSE)
}

shinyApp(ui, server)
```

</details>

https://github.com/rstudio/shiny/assets/5420529/8dfb9dff-3f11-4a82-a753-5df7be1477de

---

@kennedymwavu One temporary approach you can use is to put the data-generating logic into a `reactive()` that is bound to the `input$trigger` event. The following server logic (for your example) will cause the plot to be updated with every click:

```r
  iris_sample <- 
    reactive({
      cols <- sample(names(iris)[1:4], size = 2, replace = FALSE)
      iris[, cols]
    }) |>
    bindEvent(input$trigger, ignoreNULL = FALSE)

  output$myPlot <- renderPlot({
    cat("`renderPlot()` expr executed.\n")
    plot(
      iris_sample(),
      col = iris[["Species"]],
      pch = 19
    )
  })
```

**Comment by gadenbuie** [comment]
After thinking on this, rather than calling it a temporary solution, I'm going to fully recommend my suggestion above. Using `bindEvent()` with the `render*()` functions is supported but very nuanced. On the other hand, using `bindEvent()` with `observe()` and `reactive()` functions is much more clear and easy to reason about. So I recommend sticking to using `bindEvent()` with `reactive()` wherever possible.

The reason that `renderPlot()` behaves somewhat differently comes down to the differences between `bindEvent(observe(...), input$trigger)` and `bindEvent(reactive(...), input$trigger)`. An `observe(...)` will run the wrapped expression _every time_ `input$trigger` changes, but a `reactive(...)` only updates when `input$trigger` changes _and when_ its value is used in another reactive expression.

Because there are many ways that a plot should redraw, `renderPlot({ ... })` actually takes your expression in `{ ... }` and turns it into a `reactive()` where we add other conditions, such as the window size and other factors. When those things change, or when the reactive values used in `{ ... }` change, the plot updates. This is usually great! But `renderPlot({ ... }) |> bindEvent(input$trigger)` receives that `reactive()` object and binds it to the `input$trigger` event. The end result is that plot will only update:

1. When the value in `{ ... }` changes, or when the browser window changes
2. AND when the user clicks `input$trigger`.

This explains the behavior seen in our examples. Updating `input$trigger` is no longer enough to actually cause the plot to re-render.

So your options are:

1. Extract the logic that you care about updating with `input$trigger` into a `reactive()` and apply `bindEvent()` directly to that reactive.
2. Add `req(input$trigger)` into the `renderPlot()` expression to make the plot dependent on `input$trigger`, rather than using `bindEvent()`.

I hope this help clarify what's happening here and please don't hesitate to reach out if you have any questions.

**Comment by wch** [comment]
I agree that it is strange that the `renderPlot() |> bindEvent(input$trigger)` doesn't re-render on every click, but that other render functions, like `renderText()` do re-render, like in this example:

```R
library(shiny)

ui <- fluidPage(
  actionButton(inputId = "trigger", label = "New text"),
  textOutput("txt")
)

server <- function(input, output, session) {
  output$txt <- renderText({
    cat("`renderText()` expr executed.\n")
    paste(sample(letters, size = 10), collapse="")
  }) |> bindEvent(input$trigger)
}

shinyApp(ui, server)
```

The reason for this, as @gadenbuie mentioned, is that the user-provided expression in `renderPlot` actually gets wrapped up into a `reactive` which watches for changes in the height and width of the plot. Here's a bit more information about what's going on.

Roughly speaking, the functions are wrapped like this (note that this is a simplified explanation -- there's actually more going on):

1. At the innermost layer is the expression provided by the user in the app (which is converted to a function). I'll call this the **user code**.
2. This is wrapped by a `reactive()` which watches for changes in the height and width in the web page. When the height/width changes, it re-draws the plot, but does not re-execute the user code. This is intentional -- it improves performance, and in this app, it would prevent new random data from being generated when the window is resized.
3. The `reactive()` is in turn wrapped by the `bindEvent()`.

If you run your original example, click the button to draw the plot the first time, then resize the window, then click the button again, you will see the plot redraw, but it will not re-run the user code, and so it will display the same data that was generated the first time around.

So to sum up, the reason that `renderPlot() |> bindEvent()` behaves this way is because of the performance optimization in `renderPlot()` to efficiently handle resizing.


If you want the user code to re-execute when the `input$trigger` changes, the order of wrapping would have to be: 1. user code, 2. `bindEvent()` 3. `reactive()`. However, `bindEvent()` can't wrap a regular function directly, so you'd need to wrap the user code in a `reactive()` before passing it on to `bindEvent()`.

I agree with the options that @gadenbuie provided: Either use a separate `reactive() |> bindEvent(input$trigger)` for the user code, or use `req(input$trigger)` inside of the `renderPlot()`.

**Comment by kennedymwavu** [comment]
@gadenbuie and @wch  Thank you so much for the detailed explanations. I now have a good idea of why and what's happening. 

Would it be possible to add a short note under `?bindEvent` for posterity? The current docs on `?bindEvent` read:

> bindEvent() can be used with reactive expressions, render functions, and observers

which gives the expectation that it works well for all `render*` functions, including `renderPlot()`.

**Comment by wch** [comment]
@kennedymwavu Yes, I agree it would be good if the `bindEvent` documentation mentioned that it's a bit different for `renderPlot`.
